tests/e2e/framework/kubernetes.go

```diff
@@ -144,42 +144,57 @@ func (k *KubeInfo) Teardown() error {
 	glog.Info("Cleaning up kubeInfo")
 	var err error
 
-	if *rbacfile != "" {
+	if *skipSetup {
+		return err
+	}
 
-		testRbacYaml := filepath.Join(k.TmpDir, "yaml", filepath.Base(*rbacfile))
-		if _, err = os.Stat(testRbacYaml); os.IsNotExist(err) {
-			glog.Errorf("%s File does not exist", testRbacYaml)
-		} else if err = util.KubeDelete(k.Namespace, testRbacYaml); err != nil {
-			glog.Errorf("Rbac deletion failed, please remove stale ClusterRoleBindings")
+	istioYaml := nonAuthInstallFileNs
+	if *clusterWide {
+		if *authEnable {
+			istioYaml = authInstallFile
+		} else {
+			istioYaml = nonAuthInstallFile
+		}
+	} else {
+		if *authEnable {
+			istioYaml = authInstallFileNs
 		}
 	}
 
-	if k.namespaceCreated {
-		if err = util.DeleteNamespace(k.Namespace); err != nil {
-			glog.Errorf("Failed to delete namespace %s", k.Namespace)
-			return err
-		}
+	testIstioYaml := filepath.Join(k.TmpDir, "yaml", istioYaml)
 
-		// confirm the namespace is deleted as it will cause future creation to fail
-		maxAttempts := 15
-		namespaceDeleted := false
-		totalWait := 0
-		for attempts := 1; attempts <= maxAttempts; attempts++ {
-			namespaceDeleted, err = util.NamespaceDeleted(k.Namespace)
-			if namespaceDeleted {
-				break
-			}
-			totalWait += attempts
-			time.Sleep(time.Duration(attempts) * time.Second)
+	if *useInitializer {
+		testInitializerYAML := filepath.Join(k.TmpDir, "yaml", filepath.Base(*initializerFile))
+
+		if err := util.KubeDelete(k.Namespace, testInitializerYAML); err != nil {
+			glog.Errorf("Istio core %s deletion failed", testIstioYaml)
```

Comment:
Failure to delete the initializer and istio should return an error, otherwise we leak resources into subsequent tests.

Reply:
Yeah I think since we are still running multiple tests sequentially in the same cluster, we might want to fail at first test failure instead of continuing. You can just return the error, here and we can do the rest in another PR.

Reply:
In fact doing kube delete on the istio.yaml deletes the namespace, this is why there is no need to delete explicitely.

Reply:
The initializer and CRD definitions themselves are not namespaced. They will persist even after the istio-system namespace is deleted if not explicitly deleted.

Reply:
The namespace resource is located in the file before the other children resources, which is useful during create, but unfortunately it causes it to be deleted first, and trigger a hierarchical delete. Because of that, the command will issue some errors about not being able to delete resources, because they no longer exist. This is not a real error, this is why it should not stop.

Ideally, kube delete should use reverse order of resources in the file.

Reply:
@ayj I actually realized that since we put the CRDs in the yaml file, this will delete the non namespaced CRDs holders for the other runs. So for the shared cluster we still need to just delete the namespace.

Reply:
1. Initializer is in its own file and should only be run in non-shared test cluster. We should return an error if its deletion fails.

2. CRD probably ought to be in their own file so long as shared test clusters are used. CRD could be (re)created per test and never deleted to avoid breaking other concurrent tests in other namespaces.

3. Can the namespace creation be moved out of the common istio.yaml? Individual components could be deleted and verified that their deletion is successful as part of test.

Reply:
1. Ack the initializer

2. CRD: the issue is just for the test cluster, and I will revert my code to not do kube delete for the shared cluster, but delete the namespace like before. This is to avoid introducing an extra file and extra step in the installation.

3. Not if we want the flexibility to install in a different namespace than istio-system, plus a 1 step install. Right now, updateVersion.sh --namespace <> will produce a cluster-wide file with a different namespace. It does not work to create a ns, and apply istio.yaml -n <ns>, because the rbac permissions have to have the namespace inside them, they are not portable across namespaces.

Will do the revert and that should be fine.

Reply:
SGTM